File: Dependency Inversion Principle/DependencyInjection.swift
//
//  ViewController.swift
//  DependencyInversion
//
//  Created by paige shin on 2021/06/27.
//

import UIKit

//high level module
class ProductCatalog {
    
    let productRepository: ProductRepository
    
    init(productRepository: ProductRepository) {
        self.productRepository = productRepository
    }
    
    
    func displayProductName() {
        let names = productRepository.getListOfProductNames()
        
        //display product
        print(names)
    }
    
}

class ProductFactory {
    
    static func create() -> SQLProductRepository {
        return SQLProductRepository()
    }
    
}

//Abstraction
protocol ProductRepository {
    
    func getListOfProductNames() -> [String]
    
}

//Low Level Module
class SQLProductRepository: ProductRepository {
 
    func getListOfProductNames() -> [String] {
        return ["Hello", "World"]
    }
    
}

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        let productRepository = ProductFactory.create()
        let productDisplay = ProductCatalog(productRepository: productRepository)
        productDisplay.displayProductName()
        
    }


}

File: Dependency Inversion Principle/DependencyInjection.java
//High level module 
public class ProductCatalog {
    
    private ProductRepository productRepository;

    public ProductCatalog(ProductRepository productRepository) {
        this.productRepository = productRepository
    }
    
    public void listAllProducts() {
        List<String> listAllProducts = listAllProducts.getAllProductNames();
        //Display Products 
    }
}


public class ProductFactory {

    public static ProductRepository create() {
        return new SQLProductRepository();
    }

}

//Abstraction 
public interface ProductRepository {
    public List<String> getAllProductNames();
}

//Low Level Module 
public class SQLProductRepository implements ProductRepository {

    public List<String> getAllProductNames() {
        return ArrayList.asList("Soap", "Toothpaste")
    }

}

public class Root {

    public static void main(String[] args) {

        ProductRepository productRepository = ProductFactory.create();
        ProductCatalog ProductCatalog = new ProductCatalog(productRepository);
        ProductCatalog.listAllProducts();

    }

}